datadog_checks_base/datadog_checks/base/checks/win/wmi/sampler.py

```diff
@@ -157,15 +158,60 @@ def _query_sample_loop(self):
             self.logger.info("exception in CoInitialize: %s", e)
             raise
 
+        additional_args = []
+        if self.provider != ProviderArchitecture.DEFAULT:
+            context = Dispatch("WbemScripting.SWbemNamedValueSet")
+            context.Add("__ProviderArchitecture", self.provider)
+            additional_args = [None, "", 128, context]
+
+        locator = Dispatch("WbemScripting.SWbemLocator")
+
         while True:
             self._runSampleEvent.wait()
+            if self._stopping:
+                self.logger.debug("_query_sample_loop stopping")
+                return
+
             self._runSampleEvent.clear()
-            if self.is_raw_perf_class and not self._previous_sample:
-                self._current_sample = self._query()
 
-            self._previous_sample = self._current_sample
-            self._current_sample = self._query()
-            self._sampleCompleteEvent.set()
+            try:
+                # Create a new WMI connection.
+                self.logger.debug(
+                    u"Connecting to WMI server (host=%s, namespace=%s, provider=%s, username=%s).",
+                    self.host,
+                    self.namespace,
+                    self.provider,
+                    self.username,
+                )
+                connection = locator.ConnectServer(
+                    self.host, self.namespace, self.username, self.password, *additional_args
+                )
+
+                if self.is_raw_perf_class and not self._previous_sample:
+                    self._current_sample = self._query(connection)
+
+                self._previous_sample = self._current_sample
+                self._current_sample = self._query(connection)
+
+            except Exception as e:
+                self.logger.error("exception in _query_sample_loop: %s", e)
+            finally:
+                connection = None
+                self._sampleCompleteEvent.set()
+
+    def __enter__(self):
+        """
+        I am the context manager
+        """
+        return self
+
+    def __exit__(self, type, value, traceback):
+        """
+        Dispose of the internal thread
+        """
+        self._stopping = True
+        self._runSampleEvent.set()
```

Comment:
I'm thinking *maybe* we should wait for the thread to terminate? (We don't store the `thread` on the instance yet, but we could and then reference it here.)

```python
self.thread.join()
```

Do you think that'd make sense?

Reply:
I now wait on the complete event